src/mcelog.c

```diff
@@ -0,0 +1,431 @@
+  vl.values = &(value_t){.derive = (derive_t)mr.uncorrected_err_total};
+  plugin_dispatch_values(&vl);
+
+  return 0;
+}
+
+static void *poll_worker(__attribute__((unused)) void *arg) {
+  char buf[DATA_MAX_NAME_LEN] = {0};
+  mcelog_memory_rec_t memory_record;
+  _Bool received_data = 0;
+  FILE *p_file;
+
+  struct pollfd poll_fd = {
+      .fd = g_mcelog_config.sock_fd, .events = POLLIN | POLLPRI,
+  };
+
+  while (1) {
+
+    pthread_mutex_lock(&mcelog_thread_lock);
+    if (mcelog_thread_shutdown == 1) {
+      pthread_mutex_unlock(&mcelog_thread_lock);
+      pthread_exit(NULL);
+    }
+    pthread_mutex_unlock(&mcelog_thread_lock);
+    int poll_ret = poll(&poll_fd, 1, POLL_TIMEOUT);
+
+    if (poll_ret > 0) {
+      if (poll_fd.revents & POLLNVAL) {
+        if (connect_to_client_socket() != 0)
+          usleep(POLL_TIMEOUT);
+      }
+      if ((poll_fd.revents & POLLERR) || (poll_fd.revents & POLLHUP)) {
+        /* connection is broken */
+        shutdown(g_mcelog_config.sock_fd, 0);
+        ERROR("%s: Connection to socket is broken", MCELOG_PLUGIN);
+        notification_t n = {
+            NOTIF_FAILURE, cdtime(), "", "", MCELOG_PLUGIN, "", "", "", NULL};
+        ssnprintf(n.message, sizeof(n.message),
+                  "Connection to mcelog socket is broken.");
+        sstrncpy(n.type_instance, "mcelog_status", sizeof(n.type_instance));
+        mcelog_dispatch_notification(n);
+        break;
+      }
+      if ((poll_fd.revents & POLLIN) || (poll_fd.revents & POLLPRI)) {
+        if ((p_file = fdopen(dup(g_mcelog_config.sock_fd), "r")) != NULL) {
+          memset(buf, 0, sizeof(buf));
+          while (fgets(buf, sizeof(buf), p_file)) {
+            /* Got empty line or "done" */
+            if ((strlen(buf) == 1) ||
+                ((strlen(buf) >= 5) && (!strcmp(buf, "done\n")))) {
+              if (!received_data) {
+                break;
+              } else {
+                notification_t n = {NOTIF_OKAY,    cdtime(), "", "",
+                                    MCELOG_PLUGIN, "",       "", "",
+                                    NULL};
+                ssnprintf(n.message, sizeof(n.message),
+                          "Got memory errors info.");
+                sstrncpy(n.type_instance, "memory_erros",
+                         sizeof(n.type_instance));
+
+                if (mcelog_prepare_notification(&n, memory_record) == 0)
+                  mcelog_dispatch_notification(n);
+                if (mcelog_submit(memory_record) != 0)
+                  ERROR("%s: Failed to submit memory errors", MCELOG_PLUGIN);
+                memset(&memory_record, 0, sizeof(memory_record));
+              }
+            }
+            if (strlen(buf) >= 5) {
+              if (!strcmp(buf, "done\n")) {
+                DEBUG("%s: done", MCELOG_PLUGIN);
+                break;
+              }
+              if (!memcmp(buf, socket_str, strlen(socket_str))) {
```

Comment:
This has a high risk of reading uninitialized memory from `buf`. Please use `strncmp()` instead.